[21.05] extract osds into separate units

ctheune opened the pull request:
We had hosts run out of memory and the OSDs were all subsumed in the same cgroup which caused them to be killed together because systemd values consistency in a unit's state ...

There might be interesting cases here if NixOS decides to activate/deactivate anything, but hopefully the current setup will cause the "group unit" to catch this. I can imagine cases where this might not be the case, but I can't show that they actually exist currently.

Re PL-131646

@flyingcircusio/release-managers

## Release process

Impact: internal only

Changelog:
- move Ceph OSD processes into individual systemd service units per OSD
  - adjust `fc-ceph` accordingly
- rename service for managing all OSDs from `fc-ceph-osds.service` -> `fc-ceph-osds-all.service` 

## Security implications

- [x] [Security requirements](https://wiki.flyingcircus.io/System_Development_Guideline#Security_requirement_principles_and_testing) defined? (WHERE)
  - [x] automatic transition of running osd services at unattended upgrade time
  - [x] availability: prevent systemd from stopping all osds at the failure of a single one, just because it considers all of them to belong together
- [x] Security requirements tested? (EVIDENCE)
  - [x] automated tests still pass
  - [x] manually tested and documented proper automatic transition of running OSD services at unattended system upgrade
  - [x] extensive manual `fc-ceph` operations tests in dev cluster

ctheune committed:
fc-ceph/osds: extract OSDs into separate units to avoid shared fate

We had hosts run out of memory and the OSDs were all subsumed in the
same cgroup which caused them to be killed together because systemd
values consistency in a unit's state ...

There might be interesting cases here if NixOS decides to
activate/deactivate anything, but hopefully the current setup will cause
the "group unit" to catch this. I can imagine cases where this might
not be the case, but I can't show that they actually exist currently.

Re PL-131646
co-authored by os@flyingcircus.io

osnyx committed:
ceph osds: rename systemd service to trigger restart for migration

Before moving individual OSDs to their own units, the old `fc-ceph-osds`
unit with all osds needs to be stopped to allow the start of the
separated OSD units again.
This is achieved by renaming that service, making the NixOS
switch-to-configuration logic stop the old service and start the new
service at system update time.

This has been tested in our dev cluster:

stopping the following units: fc-blockdev.service, fc-ceph-mgr.service, fc-ceph-mon.service, fc-ceph-osds.service, fc-ceph-rgw.service, systemd-udevd-control.socket, systemd-udevd-kernel.socket, systemd-udevd.service, telegraf.service
NOT restarting the following changed units: systemd-fsck@dev-disk-by\x2dlabel-boot.service
activating the configuration...
setting up /etc...
reloading user units for os...
setting up tmpfiles
reloading the following units: dbus.service
restarting the following units: sensu-client.service
starting the following units: fc-blockdev.service, fc-ceph-mgr.service, fc-ceph-mon.service, fc-ceph-rgw.service, systemd-udevd-control.socket, systemd-udevd-kernel.socket, telegraf.service
the following new units were started: fc-ceph-osd@0.service, fc-ceph-osd@16.service, fc-ceph-osd@20.service, fc-ceph-osd@4.service, fc-ceph-osd@6.service, fc-ceph-osd@8.service, fc-ceph-osds-all.service, fc-update-channel.timer

osnyx committed:
ceph osds: pull in fc-blockdev via unit dependency instead of manual start

osnyx committed:
fc-ceph osd: minor cleanups

osnyx reviewed (approved):
LGTM now, but to be fair I did most of the fine-tuning and adjustments of the core approach, so someone else needs to review as well.

osnyx committed:
ceph luminous and nautilus: remove nix expressions as well

follow-up to 5d9e6562169a409127fa1d70ad136656b3235b78 where both packages
had only been removed as references from the overlay, but not as files
from the repo
